tests/integration/utils.py

```diff
@@ -463,3 +472,165 @@ def _validate_expected_dep_file_contents(dep_contents_file: Path, output_dir: Pa
         dep_file = output_dir / "deps" / path
         assert dep_file.exists()
         assert dep_file.read_text() == expected_content
+
+
+def retrieve_changed_files_from_git() -> tuple[Path, ...]:
+    repo = Repo(".", search_parent_directories=True)
+    try:
+        # HEAD would be correctly defined locally
+        files = repo.git.diff("--name-only", f"{repo.active_branch.name}..main").split("\n")
+    except TypeError:  # will be raised when in detached HEAD state, e.g. in Github CI
```

Comment:
Not opposing per se, but can we trust it in general? IOW would it only happen in GH CI? Can't this happen in local envs is what I'm asking? For some reason it's making me uneasy.
Wouldn't 
```
if not(os.getenv("CI") or os.getenv("GITHUB_WORKSPACE")):
  repo.commit("HEAD")
else:
  repo.git.diff(...)
```
be slightly more bulletproof?

Based on https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/store-information-in-variables#default-environment-variables which are always set in GH CI env?

Reply:
Uhm, lol, this just happened to me as I added commits on top of your PR locally (I swear I didn't do it on purpose :D ). Whenever I fetch PRs locally I set git to always put them in a detached HEAD state rather than a specific branch (because once I'm done with the code review I just checkout elsewhere and don't need to delete a branch I may not need again) and I was wondering what was wrong with the test results :D.

Reply:
I probably shouldn't have followed with my own example, because it steered the conversation in a way I didn't intend to.

My main point was that it behaves somewhat unexpectedly if it happens, solely because you make the assumption that GitHub CI is the only env which will use detached HEADs which is a standard git feature. Moreover, you also had to add a commentary to explain the exception instead of going the env variable approach where it's immediately clear which case it handles. That is not to say that someone cannot locally set all those GH variables explicitly, but who'd do that honestly voluntarily?

Reply:
> In this specific case one could trim the scope of testing further with -k test_pattern if needed (I checked that). If, however, someone wants to run unrelated tests on a commit that does not affect tested code then these tests will be skipped with correct reason. I.e. if I detach HEAD at https://github.com/containerbuildsystem/cachi2/commit/45c8c7c9b12fb0158f623f41c13f090ca218d513 and then run tox -eintegration -- -k test_gomod_packages I'll end up with 12 skipped and 40 deselected tests vs 51 skipped for just tox -eintegration. 

This is all nice and I don't disagree, but I'm more after the automatic behaviour without any CLI args.

Reply:
I mentioned CLI args only to highlight that if someone is up to some complicated workflow they could still have it. The whole point of the change to be drop-in compatible with any regular workflow and to save time.